### ISSUE TITLE
Directory expansion in map_each can't handle nested directory values

### ISSUE THREAD

**Issue by jkrems** [bazelbuild/bazel#13730]
### Description of the problem / feature request:

We'd like to access analysis time values available in a rule implementation function, from within a `map_each` function.

### Feature requests: what underlying problem are you trying to solve with this feature?

We'd like to use a primitive value from the rule implementation inside of the argument mapper (stripping a target-specific prefix). There's currently only one way to do that afaik:

```bzl
def _format_arg(pair, expander):
    [
        "%s:%s" % (pair.s, file.short_path)
        for file in expander.expand(pair.d)
    ]

def _my_rule_impl(ctx):
    args = ctx.actions.args()
    args.add_all([struct(d = d, s = ctx.attr.str) for d in ctx.files.data], map_each = _format_arg)
    ctx.actions.write(ctx.declare_file(ctx.label.name), args)

my_rule = rule(
    implementation = _my_rule_impl,
    attrs = {
        "str": attr.string(),
        "data": attr.label_list(allow_files = 1),
    },
)
```

The above is a simplified example and already a bit hard to follow - and it's actually subtly broken. The directory expander will never really expand directories because it's using an artifact expander that doesn't know about those artifacts. Instead the expander will silently return the directory artifact, leading to broken output.

What we'd really want to write is something like the following:

```bzl
def _my_rule_impl(ctx):
    def _format_arg(file):
        "%s:%s" % (ctx.attr.str, file.short_path)

    args = ctx.actions.args()
    args.add_all(ctx.files.data, map_each = _format_arg)
    ctx.actions.write(ctx.declare_file(ctx.label.name), args)
# [...]
```

That's prevented by the enforcement of top-level functions for `map_each`. There's [`partial.make` in skylib](https://github.com/bazelbuild/bazel-skylib/blob/main/docs/partial_doc.md) but it's not something you can pass into `map_each`. Otherwise we'd also be more than happy to write (explicitly closing over the value):

```bzl
def _format_arg(str, file):
    "%s:%s" % (str, file.short_path)

def _my_rule_impl(ctx):

    args = ctx.actions.args()
    args.add_all(ctx.files.data, map_each = partial.make(_format_arg, ctx.attr.str))
    ctx.actions.write(ctx.declare_file(ctx.label.name), args)
# [...]
```

#### Honorable Mentions

Other approaches we considered (H/T @soldair):

* Add a new `add_all` parameter (~`map_each_clientdata`) that will be forwarded to each call to `map_each` as a third argument.
* Add a new `add_all` parameter (~`map_each_inputs`) that ensures that the directory expander consistently knows about the relevant artifacts, even when they are wrapped in some data structure.

### What operating system are you running Bazel on?

Linux

### What's the output of `bazel info release`?

n/a

###  Have you found anything relevant by searching the web?

* https://github.com/bazelbuild/bazel/issues/11811

### Any other information, logs, or outputs that you want to share?

n/a

**Comment by benjaminp** [comment]
There is additional context in #12701.

**Comment by jkrems** [comment]
Thanks for the link! What I'd like to prevent is that we effectively retain more memory (one struct per file etc.) which is the current workaround for the lack of partial application or closure state. But it sounds like `clientdata` was mentioned and there didn't seem to be a clear objections to that approach. Which would give us the general advantage of partial application without the lower level feature in this case.

**Comment by aiuto** [comment]
@comius You might find this one interesting.

**Comment by comius** [comment]
I wonder if the particular example can be handled with additional `before_each` and `format_each` parameters? Combination of those parameters with `map_each` should already be general enough and prevent any unintended memory retention.

**Comment by aiuto** [comment]
While we are talking about improving args.add_*, I would like to fix the behavior of `format_joined`. See #13800

**Comment by jkrems** [comment]
> I wonder if the particular example can be handled with additional before_each and format_each parameters?

To clarify the proposal here, something like this?

```
def _map_arg(f):
  return f.short_path

# [...]
args.add_all(ctx.files.data, map_each = _map_arg, format_each = (ctx.attr.str + ":%s"))
```

Guess I simplified the example too much. In the real code, we're stripping path prefixes (conditionally) based on `ctx.attr.str`. So we wouldn't be able to generate a "short path" inside of `_map_arg` the way it would appear in the final result. But thanks for bringing up `format_each`! I think we have some cases where it might be sufficient.

**Comment by comius** [comment]
To me it makes sense to optionally disable "the enforcement of top-level functions for map_each". I would accept a PR that adds an `allow_closure` parameter, defaulting to False.

This way if the user is aware of possible regression, the closure may still be used.

**Comment by aiuto** [comment]
I think the `clientdata` idea would serve a lot of people with minimal effort.

**Comment by jkrems** [comment]
> To me the most straightforward thing to do is to add a new arg to add_all of additional directories to make expandable. This avoids the need to change the map_each restriction. It's a shame that the API is growing more complicated bit by bit though.

The issue with that is that it forces the creation of more complex data structures (and potentially expanding depsets) that then need to be retained into the execution phase. Which seems worse than allowing closures. Unless I'm missing something about that approach?

One thing that @tjgq brought up elsewhere is that we could add some validation on the `freevars` of the starlark function to make sure that it isn't closing over "dangerous" or expensive objects (e.g. only allowing primitive values). That might even be a path towards making ~`allow_closure` the default behavior (with those restrictions in place).

**Comment by DhanalakshmiDurairaj** [comment]
Hi
Greetings to all
list DirectoryExpander.expand(file)

Please help me to get an understanding on DirectoryExpander API which is specified in https://bazel.build/rules/lib/builtins/DirectoryExpander.

How to load this API, to expand directory and get list of files. I have to expand directory and get all the files inside the directory into List variable. Pls help me on this.

Thanks & Regards,
D.Dhanalakshmi

**Comment by brandjon** [comment]
Please see the design discussion in #18825. I'm revisiting this issue, and I aim to eliminate `allow_closure` in favor of clientdata. This avoids correctness bugs in digesting non-global Starlark functions.
